java/google/registry/rde/imports/RdeDomainImportAction.java

```diff
@@ -0,0 +1,164 @@
+        this.importFileName);
+    response.sendJavaScriptRedirect(createJobPath(mrRunner
+        .setJobName("Import domains from escrow file")
+        .setModuleName("backend")
+        .runMapOnly(
+            createMapper(),
+            ImmutableList.of(createInput()))));
+  }
+
+  /**
+   * Creates a new {@link RdeDomainInput}
+   */
+  private RdeDomainInput createInput() {
+    return new RdeDomainInput(mapShards, importBucketName, importFileName);
+  }
+
+  /**
+   * Creates a new {@link RdeDomainImportMapper}
+   */
+  private RdeDomainImportMapper createMapper() {
+    return new RdeDomainImportMapper(importBucketName);
+  }
+
+  /** Mapper to import domains from an escrow file. */
+  public static class RdeDomainImportMapper extends Mapper<JaxbFragment<XjcRdeDomainElement>, Void, Void> {
+
+    private static final long serialVersionUID = -7645091075256589374L;
+    private final String importBucketName;
+    private transient RdeImportUtils importUtils;
+
+    public RdeDomainImportMapper(String importBucketName) {
+      this.importBucketName = importBucketName;
+    }
+
+    private RdeImportUtils getImportUtils() {
+      if (importUtils == null) {
+        importUtils = createRdeImportUtils();
+      }
+      return importUtils;
+    }
+
+    /**
+     * Creates a new instance of RdeImportUtils.
+     */
+    private RdeImportUtils createRdeImportUtils() {
+      return new RdeImportUtils(
+          ofy(),
+          new SystemClock(),
+          importBucketName,
+          new GcsUtils(GCS_SERVICE, ConfigModule.provideGcsBufferSize()));
+    }
+
+    @Override
+    public void map(JaxbFragment<XjcRdeDomainElement> fragment) {
+      final XjcRdeDomain xjcDomain = fragment.getInstance().getValue();
+      try {
+        logger.infofmt("Converting xml for domain %s", xjcDomain.getName());
+        // Record number of attempted map operations
+        getContext().incrementCounter("domain imports attempted");
+        ofy().transact(new VoidWork() {
+          @Override
+          public void vrun() {
+            DomainResource domain =
+                XjcToDomainResourceConverter.convertDomain(xjcDomain);
+            logger.infofmt("Saving domain %s", domain.getFullyQualifiedDomainName());
+            getImportUtils().importDomain(domain);
+            logger.infofmt("Domain %s was imported successfully",
+                domain.getFullyQualifiedDomainName());
+          }
+        });
+        getContext().incrementCounter("domains saved");
+      } catch (ResourceExistsException e) {
+        // Record the number of domains already in the registry
+        getContext().incrementCounter("domains skipped");
+        logger.infofmt("Domain %s already exists", xjcDomain.getName());
+      } catch (Exception e) {
+        getContext().incrementCounter("domain import errors");
+        logger.severefmt(e, "Error processing domain %s; xml=%s", xjcDomain.getName(), xjcDomain);
+        throw new RuntimeException(e);
```

Comment:
Ok, sorry, I didn't notice the log here. If you are logging, this is the log-and-throw antipattern. It would be better to throw a new exception always with this message:

class DomainImportException extends RuntimeException() {
DomainImportException(String message, Throwable cause) {
super(message, cause);
}
}
throw new DomainImportException(String.format("Domain %s already exists", xjcDomain.getName());